### PR TITLE
Issue #634: satelite projects should use latest checkstyle as soon at is appears in maven repo

### LINKED ISSUE #634: satelite projects should use latest checkstyle as soon at is appears in maven repo
Right now we do bump a version in a few projects https://github.com/checkstyle/checkstyle/wiki/How-to-make-a-release#update-for-released-version-reference right after release.

Let's find a way to define in pom.xml the latest snapshot and most recent version. And the latest released version when we define dependencies. To let us skip such updates at all.

CI of such projects should be able to get the latest version from https://repo1.maven.org/maven2/com/puppycrawl/tools/checkstyle/maven-metadata.xml and run projects with such version.

We can have in all such repos, a GitHub action that runs once a day check https://repo1.maven.org/maven2/com/puppycrawl/tools/checkstyle/maven-metadata.xml, put such values in required files, and send PRs with changes (PR is necessary as we need notice is any problems happens if we introduce breaking changes)

TASK:

- [ ] Version Bump Script in `releasenotes-builder/pom.xml`
- [ ]  Version Bump Script in `patch-diff-report-tool/pom.xml`
- [ ]  GitHub Action to run both the scripts.

### PR DESCRIPTION
**Pull request by Rahulkhinchi03**
Partially Fixes #634 

Testing: 
When version is different : [ACTION](https://github.com/Rahulkhinchi03/contribution/runs/7939630676?check_suite_focus=true)
When a version is the same: [ACTION](https://github.com/Rahulkhinchi03/contribution/runs/7939635057?check_suite_focus=true)

### CODE REVIEW

File: .github/workflows/bump_project_version.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+name: "Bump Project Version"
+on:
+    schedule:
+        - cron: '0 0 * * *'
+permissions:
+    contents: write
+    pull-requests: write
+jobs:
+    bump:
+        name: Bump Project Version
```
Comment (romani) [inline .github/workflows/bump_project_version.yml:16]: Bump CS version in repo

File: .github/workflows/bump_project_version.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        runs-on: ubuntu-latest
+        steps:
+            - name: Run Shell Script
+              run: |
+                  ./.ci/bump-patch-tool-version.sh
+                  ./.ci/bump-releasenotes-builder-version.sh
```
Comment (romani) [inline .github/workflows/bump_project_version.yml:22]: see https://github.com/checkstyle/contribution/pull/635#pullrequestreview-1079599409

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: .github/workflows/bump_project_version.yml (hunk 1 above)
Comment (Rahulkhinchi03) [reply .github/workflows/bump_project_version.yml]: Changes Done

File: .github/workflows/bump_project_version.yml (hunk 2 above)
Comment (Rahulkhinchi03) [reply .github/workflows/bump_project_version.yml]: Changes Done

File: .github/workflows/bump_project_version.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
+#############################################################################
+# GitHub Action to Bump Project Version.
```
Comment (romani) [inline .github/workflows/bump_project_version.yml:2]: bump_project_version.yml -> bump-cs-version.yml 

>GitHub Action to Bump Project Version.

GitHub Action to Bump Checkstyle Version in repository.

File: .github/workflows/bump_project_version.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+#
+#############################################################################
+name: "Bump CS Version in REPO"
+on:
+    schedule:
+        - cron: '0 0 * * *'
+    workflow_dispatch:
+permissions:
+    contents: write
+    pull-requests: write
```
Comment (romani) [inline .github/workflows/bump_project_version.yml:14]: why you need write for PRs ?

File: .github/workflows/bump_project_version.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    bump:
+        name: Bump CS version in REPO
+        runs-on: ubuntu-latest
+        steps:
+            - name: Checkout the latest code
+              uses: actions/checkout@v2
+              with:
+                  token: ${{ secrets.GITHUB_TOKEN }}
+            - name: Get Version
+              id: VERSION
```
Comment (romani) [inline .github/workflows/bump_project_version.yml:25]: why you need this id ?

File: .github/workflows/bump_project_version.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+                  ./.ci/bump-cs-version-in-patch-diff-report-tool.sh ${{ steps.VERSION.outputs.LATEST_VERSION }}
+                  ./.ci/bump-cs-version-in-releasenotes-builder.sh ${{ steps.VERSION.outputs.LATEST_VERSION }}
+            - name: Check Changes
+              run: |
+                  git diff
+            - name: Commit and Push Changes
+              run: |
+                  git config --global user.name 'github-actions[bot]'
+                  git config --global user.email 'github-actions[bot]@users.noreply.github.com'
+                  git commit -am "minor: Bump project version to ${{ steps.VERSION.outputs.LATEST_VERSION }}"
```
Comment (romani) [inline .github/workflows/bump_project_version.yml:41]: you need to protect creation of commit without changes, this line will  fail if if changes detected.
https://github.com/checkstyle/checkstyle/blob/a97c3e8196c90fd8cf90a755f0b7ef5fd2824d29/.ci/validation.sh#L1008-L1010

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: .github/workflows/bump_project_version.yml (hunk 5 above)
Comment (Rahulkhinchi03) [reply .github/workflows/bump_project_version.yml:25]: This id is needed to set `VAR` so we can use it as latest version and pass it as argument

File: .github/workflows/bump_project_version.yml (hunk 3 above)
Comment (Rahulkhinchi03) [reply .github/workflows/bump_project_version.yml]: Changes Done

File: .github/workflows/bump_project_version.yml (hunk 4 above)
Comment (Rahulkhinchi03) [reply .github/workflows/bump_project_version.yml]: Changes Done. No Need

File: .github/workflows/bump_project_version.yml (hunk 6 above)
Comment (Rahulkhinchi03) [reply .github/workflows/bump_project_version.yml:39]: Changes Done

File: .github/workflows/bump_project_version.yml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+                  LATEST_VERSION=$(curl $MAVEN_METADATA | grep -oPm1 "(?<=<latest>)[^<]+")
+                  echo "::set-output name=LATEST_VERSION::$LATEST_VERSION"
+            - name: Run Shell Script
+              run: |
+                  ./.ci/bump-cs-version-in-patch-diff-report-tool.sh ${{ steps.VERSION.outputs.LATEST_VERSION }}
+                  ./.ci/bump-cs-version-in-releasenotes-builder.sh ${{ steps.VERSION.outputs.LATEST_VERSION }}
+            - name: Commit and Push
+              run: |
+                  if [ "$(git status | grep 'Changes not staged\|Untracked files')" ]; then
+                  git diff
```
Comment (romani) [inline .github/workflows/bump_project_version.yml:36]: +2 for Indentation

Please always respect indentation for inner block or wrapping

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: .github/workflows/bump_project_version.yml (hunk 7 above)
Comment (Rahulkhinchi03) [reply .github/workflows/bump_project_version.yml:36]: I am sorry that you always have to me tell me that :(

**Review by romani — APPROVED** [review]
Ok to merge

**Pull request merged**